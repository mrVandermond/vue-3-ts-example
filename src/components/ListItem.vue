<template>
  <div
    :class="{
      internal: props.isInternal,
      'is-open': isOpen
    }"
    class="list-item"
  >
    <div
      class="list-item-header"
      @click="onClickHeader"
    >
      <ListItemButton
        :is-open="isOpen"
        class="list-item-button"
      />

      <slot name="label" />
    </div>

    <div
      :class="{
        'is-loading': props.isLoadingContent,
      }"
      class="list-item-content"
    >
      <Loader v-if="props.isLoadingContent" />

      <Error
        v-else-if="!props.isLoadingContent && props.isFailLoadingContent"
        :is-title-from-right="true"
      />

      <slot
        v-else
        name="content"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue';
import ListItemButton from '@/components/ListItemButton.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';

const isOpen = ref(false);

const props = withDefaults(defineProps<{
  isInternal?: boolean;
  isLoadingContent?: boolean;
  isFailLoadingContent?: boolean;
}>(), {
  isInternal: false,
  isLoadingContent: false,
  isFailLoadingContent: false,
});
const emit = defineEmits<{
  (event: 'click', isOpen: boolean): void;
}>();

const onClickHeader = (): void => {
  isOpen.value = !isOpen.value;
  emit('click', isOpen.value);
};
</script>

<style lang="scss">
.list-item {
  padding: 24px 0;

  &.internal {
    margin-left: 56px;
  }

  &.is-open {
    padding-bottom: 0;
  }
}
.list-item-button {
  margin-right: 24px;
}
.list-item-header {
  cursor: pointer;
  display: flex;
  align-items: center;
}
.list-item-content.is-loading {
  min-height: 320px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
}
</style>
