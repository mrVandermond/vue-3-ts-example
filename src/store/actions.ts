import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
