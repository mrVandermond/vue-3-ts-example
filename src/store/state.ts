import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';

export interface IState {
  favoritePhotos: IPhotoWithIsFavorite[];
  modal: IModalConfig;
  title: ITitleConfig;
}

export const state = (): IState => ({
  favoritePhotos: [],
  modal: {
    isOpen: false,
    url: '',
  },
  title: {
    isShow: false,
    text: '',
    xCoord: 0,
    yCoord: 0,
  },
});
