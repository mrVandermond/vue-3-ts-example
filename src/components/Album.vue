<template>
  <ListItem
    :is-loading-content="isLoadingPhotos"
    :is-fail-loading-content="isFailLoadingPhotos"
    is-internal
    @click="onClickItem"
  >
    <template #label>
      <div class="album-label">
        {{ props.album.title }}
      </div>
    </template>

    <template
      v-if="isOpenAlbum"
      #content
    >
      <div
        class="list-photo"
        @mousemove.passive="titleOptions.onMouseMove"
      >
        <Photo
          v-for="photo in photos"
          :key="photo.id"
          :photo="photo"
          @added-to-favorites="onAddedToFavorites"
          @remove-from-favorites="onRemoveFromFavorites"
        />
      </div>
    </template>
  </ListItem>
</template>

<script lang="ts" setup>
import type { IAlbum, IPhotoWithIsFavorite } from '@/types';
import ListItem from '@/components/ListItem.vue';
import Photo from '@/components/Photo.vue';
import { ref, watchEffect } from 'vue';
import { storeTypes } from '@/store/types';
import { customUseStore } from '@/store';
import { useTitle } from '@/composables';

const props = defineProps<{
  album: IAlbum;
}>();

const photos = ref<IPhotoWithIsFavorite[]>([]);
const isLoadingPhotos = ref(false);
const isFailLoadingPhotos = ref(false);
const isOpenAlbum = ref(false);

const store = customUseStore();
const titleOptions = useTitle();

watchEffect(() => {
  store.commit(storeTypes.EMutations.SET_TITLE_CONFIG, {
    isShow: titleOptions.isShowTitle.value,
    xCoord: titleOptions.xCoordMouse.value,
    yCoord: titleOptions.yCoordMouse.value,
    text: titleOptions.titleText.value,
  });
});

function toggleIsFavorite(index: number): void {
  photos.value[index] = {
    ...photos.value[index],
    isFavorite: !photos.value[index].isFavorite,
  };
}

async function onClickItem(isOpen: boolean): Promise<void> {
  isOpenAlbum.value = isOpen;

  if (!isOpen) return;

  isLoadingPhotos.value = true;

  try {
    const photosWithoutIsFavorite = await store.dispatch(storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID, props.album.id);

    photos.value = photosWithoutIsFavorite.map((photo) => ({
      ...photo,
      isFavorite: store.getters[storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](photo.id),
    }));
  } catch (error) {
    isFailLoadingPhotos.value = true;
  }

  isLoadingPhotos.value = false;
}
function onAddedToFavorites(photo: IPhotoWithIsFavorite): void {
  store.commit(storeTypes.EMutations.ADD_TO_FAVORITES, [photo]);

  const photoIndex = photos.value.findIndex((item) => item.id === photo.id);

  if (photoIndex === -1) return;

  toggleIsFavorite(photoIndex);
}
function onRemoveFromFavorites(photo: IPhotoWithIsFavorite): void {
  store.commit(storeTypes.EMutations.REMOVE_FROM_FAVORITES, photo);

  const photoIndex = photos.value.findIndex((item) => item.id === photo.id);

  if (photoIndex === -1) return;

  toggleIsFavorite(photoIndex);
}
</script>

<style lang="scss">
.album-label {
  font-size: 18px;
}
.list-photo {
  display: flex;
  flex-wrap: wrap;
  gap: 42px;
  padding: 32px 105px 32px 49px;
}
</style>
