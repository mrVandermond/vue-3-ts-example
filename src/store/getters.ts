import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
