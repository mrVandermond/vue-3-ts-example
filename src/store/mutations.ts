import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
