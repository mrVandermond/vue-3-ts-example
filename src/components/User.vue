<template>
  <ListItem
    :is-loading-content="isLoadingAlbums"
    :is-fail-loading-content="isFailLoadingAlbums"
    @click="onClickItem"
  >
    <template #label>
      <div class="user-label">
        {{ props.user.name }}
      </div>
    </template>

    <template
      v-if="isOpenUser"
      #content
    >
      <Album
        v-for="album in albums"
        :key="album.id"
        :album="album"
      />
    </template>
  </ListItem>
</template>

<script lang="ts" setup>
import ListItem from '@/components/ListItem.vue';
import Album from '@/components/Album.vue';
import type { IAlbum, IUser } from '@/types';
import { ref } from 'vue';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const props = defineProps<{
  user: IUser;
}>();
const store = customUseStore();
const albums = ref<IAlbum[]>([]);
const isOpenUser = ref(false);
const isLoadingAlbums = ref(false);
const isFailLoadingAlbums = ref(false);

const onClickItem = async (isOpen: boolean): Promise<void> => {
  isOpenUser.value = isOpen;

  if (!isOpen) return;

  isLoadingAlbums.value = true;

  try {
    albums.value = await store.dispatch(storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID, props.user.id);
  } catch (error) {
    isFailLoadingAlbums.value = true;
  }

  isLoadingAlbums.value = false;
};
</script>

<style lang="scss">
.user-label {
  font-weight: 500;
  font-size: 22px;
  height: 22px;
}
</style>
