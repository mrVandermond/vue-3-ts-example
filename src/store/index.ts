import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
