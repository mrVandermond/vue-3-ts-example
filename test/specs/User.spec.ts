import type { VueWrapper } from '@vue/test-utils';
import { flushPromises } from '@vue/test-utils';

import User from '@/components/User.vue';
import ListItem from '@/components/ListItem.vue';
import Album from '@/components/Album.vue';

import type { Store } from '@/store/types/overrides';
import type { IUser } from '@/types';

import { customCreateStore, key } from '@/store';
import { storeTypes } from '@/store/types';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';

import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

const user: IUser = {
  id: 1,
  name: 'Leanne Graham',
  username: 'Bret',
  email: 'Sincere@april.biz',
  address: {
    street: 'Kulas Light',
    suite: 'Apt. 556',
    city: 'Gwenborough',
    zipcode: '92998-3874',
  },
  phone: '1-770-736-8031 x56442',
  website: 'hildegard.org',
  company: {
    name: 'Romaguera-Crona',
    catchPhrase: 'Multi-layered client-server neural-net',
    bs: 'harness real-time e-markets',
  },
};

describe('User component', () => {
  let store!: Store;

  const getWrapper = (propsUser: IUser = user): TWrapperFactoryReturnType => wrapperFactory(User, {
    props: {
      user: propsUser,
    },
    global: {
      plugins: [[store, key]],
      stubs: {
        ListItem,
        Album,
      },
    },
  });

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });

    jest.spyOn(store, 'dispatch').mockImplementation(() => new Promise((resolve) => {
      resolve(Promise.resolve());
    }));
  });

  it('Рендеринг по дефолту', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('Открытие юзера', () => {
    let wrapper: TWrapperFactoryReturnType;
    let listItem: VueWrapper;

    const openListItem = async (userWrapper: TWrapperFactoryReturnType): Promise<void> => {
      listItem = userWrapper.findComponent(ListItem);

      listItem.vm.$emit('click', true);

      await flushPromises();
    };

    beforeEach(() => {
      wrapper = getWrapper();
    });

    it('Запрос прошел', async () => {
      await openListItem(wrapper);

      expect(store.dispatch).toHaveBeenCalledWith(storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID, 1);
      expect(listItem.props().isLoadingContent).toEqual(false);
      expect(listItem.props().isFailLoadingContent).toEqual(false);
    });

    it('Запрос загружается', async () => {
      jest.spyOn(store, 'dispatch').mockImplementation(() => new Promise((resolve) => {
        setTimeout(resolve, 10000);
      }));

      await openListItem(wrapper);

      expect(store.dispatch).toHaveBeenCalledWith(storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID, 1);
      expect(listItem.props().isLoadingContent).toEqual(true);
      expect(listItem.props().isFailLoadingContent).toEqual(false);
    });

    it('Запрос упал', async () => {
      jest.spyOn(store, 'dispatch').mockImplementation(() => new Promise((resolve, reject) => {
        reject(new Error('Ошибка'));
      }));

      await openListItem(wrapper);

      expect(store.dispatch).toHaveBeenCalledWith(storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID, 1);
      expect(listItem.props().isLoadingContent).toEqual(false);
      expect(listItem.props().isFailLoadingContent).toEqual(true);
    });
  });

  it('Отображение, при закрытии юзера', async () => {
    const wrapper = getWrapper();

    const listItem = wrapper.find('.user-label');

    await listItem.trigger('click');

    await listItem.trigger('click');

    expect(store.dispatch).toHaveBeenCalledTimes(1);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
