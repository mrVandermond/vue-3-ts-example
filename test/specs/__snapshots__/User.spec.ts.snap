// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component Отображение, при закрытии юзера 1`] = `
<div class="list-item">
  <div class="list-item-header">
    <list-item-button-stub isopen="false"
                           class="list-item-button"
    >
    </list-item-button-stub>
    <div class="user-label">
      Leanne Graham
    </div>
  </div>
  <div class="list-item-content">
  </div>
</div>
`;

exports[`User component Рендеринг по дефолту 1`] = `
<div class="list-item">
  <div class="list-item-header">
    <list-item-button-stub isopen="false"
                           class="list-item-button"
    >
    </list-item-button-stub>
    <div class="user-label">
      Leanne Graham
    </div>
  </div>
  <div class="list-item-content">
  </div>
</div>
`;
